src/shogun/multiclass/tree/CARTree.cpp

```diff
@@ -281,10 +281,30 @@ bool CARTree::train_machine(std::shared_ptr<Features> data)
 	{
 		prune_by_cross_validation(dense_features,m_folds);
 	}
-
 	return true;
 }
 
+void CARTree::compute_feature_importance(std::shared_ptr<bnode_t> node)
```

Comment:
you should pass the argument as `const std::shared_ptr<bnode_t>& node`

Reply:
i thought the size of shared_ptr is the same as the raw pointer, but const std::shared_ptr<bnode_t>& node is surely better :)

Reply:
hmm it's not about the size. A shared_ptr has a reference counter, and if you copy it it will do an atomic operation to increment the count, which is very slow. So if you pass it as const reference you are just passing a reference, so no copy occurs, and the const makes sure you don't modify the reference (more or less).
A raw pointer is just a memory address (more or less).